ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryRegistryImpl.java

```diff
@@ -251,6 +251,9 @@ private void registerQuery(
       final QueryMetadata query,
       final boolean createAsQuery
   ) {
+    // First initialize the query.  Once it's added to these maps/sets, it's exposed to other
+    // threads and it's important that it's initialized.
+    query.initialize();
```

Comment:
this code is brittle and I'm a little nervous about changing it around. c.f. https://github.com/confluentinc/ksql/pull/7617

@rodesai can you review this PR?

Reply:
I tried this :). We can't initialize it here without some additional changes. In the upgrade case the previous query would still be running, and would have a lock on the state store dirs. This initialize call would throw because it can't acquire the lock. We could refactor to:

1. if the query already exists in the map, stop and unregister it
2. initialize this query
3. register this query

Alternatively we could refactor QueryMetadata to be able to deal with an uninitialized Kafka Streams, and do the initialization transparently when the query is started.

Reply:
I definitely agree that this contract to the caller that the `QueryMetadata` is invalid until `initialize` is called is brittle and the ideal solution would be to make the contract easier to deal with and hopefully not throw NPEs if you violate the contract.

I was looking into `QueryMetadataImpl` and it seems to assume that it's initialized once and then is immutable (and therefore threadsafe) because there's no synchronization and fields are largely not updated.  The fact that there's a reset call on `PersistentQueryMetadataImpl` which resets kafkaStreams is a little worrying though it only seems to be called from tests.  Also, fields like `closed` and `everStarted` are not volatile as they likely should be.  (Am I right here?  Happy to make that fix since I'm already looking here.)

I could make the changes you suggest, though allowing for the object to be mutable after it's exposed publicly would require adding a lot of synchronization and complicating `QueryMetadataImpl` a bit.  I'm fine to do that, but wanted to keep things simple to target this race condition, so did your first suggestion.